warehouse/query-core/src/main/java/datawave/core/iterators/FieldExpansionIterator.java

```diff
@@ -0,0 +1,152 @@
+ */
+public class FieldExpansionIterator extends SeekingFilter implements OptionDescriber {
+
+    private static final Logger log = LoggerFactory.getLogger(FieldExpansionIterator.class);
+
+    public static final String START_DATE = "start.date";
+    public static final String END_DATE = "end.date";
+    public static final String FIELDS = "fields";
+    public static final String DATATYPES = "dts";
+
+    // required
+    private String startDate;
+    private String endDate;
+
+    private Set<String> fields;
+    private Set<String> datatypes;
+
+    // track which fields this iterator has seen and returned. this collection is not persisted between teardown and rebuilds, so unique return values
+    // are only guaranteed within a single non-interrupted scan session
+    private final Set<String> found = new HashSet<>();
+
+    private final ShardIndexKey parser = new ShardIndexKey();
+
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+        if (!validateOptions(options)) {
+            throw new IllegalStateException("FieldExpansionIterator not configured with correct options");
+        }
+
+        if (options.containsKey(FIELDS)) {
+            fields = new HashSet<>(Splitter.on(',').splitToList(options.get(FIELDS)));
+        }
+
+        if (options.containsKey(DATATYPES)) {
+            datatypes = new HashSet<>(Splitter.on(',').splitToList(options.get(DATATYPES)));
+        }
+
+        startDate = options.get(START_DATE);
+        endDate = options.get(END_DATE) + Constants.MAX_UNICODE_STRING;
+
+        super.init(source, options, env);
+    }
+
+    @Override
+    public FilterResult filter(Key k, Value v) {
+
+        // keep it simple for now
+        log.trace("tk: {}", k.toStringNoTime());
+
+        parser.parse(k);
+
+        // if field does not match, skip to next field
+        if ((fields != null && !fields.contains(parser.getField())) || found.contains(parser.getField())) {
+            log.trace("field not in set of expansion fields, or already seen this field. advancing to next field");
+            return new FilterResult(false, AdvanceResult.NEXT_CF);
+        }
+
+        // ensure key falls within the date range
+        String date = parser.getShard();
+        if (date.compareTo(startDate) < 0) {
+            // advance to start date
+            log.trace("Key before start date: {} < {}", date, startDate);
+            return new FilterResult(false, AdvanceResult.NEXT);
+        }
+
+        if (date.compareTo(endDate) > 0) {
+            // advance to next field
+            log.trace("Key after end date: {} > {}", date, endDate);
+            return new FilterResult(false, AdvanceResult.NEXT_CF);
+        }
+
+        if (datatypes != null && !datatypes.contains(parser.getDatatype())) {
+            // figure out if we are before the first datatype -- seek to first datatype
```

Comment:
This comment does not seem relevant here

Reply:
can we seek to the next datatype?

Reply:
Due to the structure of the shard index keys a seek based on the datatype is never practical. I'll correct the comment.

Reply:
Revisited this. If any datatype in the filter sorts AFTER the current datatype then we can advance via a next call. If no datatype in the filter sorts after the current datatype then no match is possible and the iterator can advance to the next field.